USF-1657: Metadata and SEO (#81)
commerce-docs/microsite-commerce-storefront: src/content/docs/implementation/marketing/metadata.mdx

---
title: Indexing and metadata
description: Boost your Adobe Commerce SEO on Edge Delivery Services! Follow our guidelines on data layers, canonical URLs, redirects, sitemaps, and metadata.
---

import Aside from '@components/Aside.astro';
import { Tabs, TabItem, Steps } from '@astrojs/starlight/components';

> This page provides guidelines for optimizing Adobe Commerce storefronts on Edge Delivery Services for SEO. It covers general recommendations for indexing and metadata.

Reviewer: I don't like leading off with sentences like "This page/This section...", etc. If you could find a way to rewrite the above section, or the one for Indexing, that'd be great.

### Folder mapping

When using [folder mapping](https://www.aem.live/developer/folder-mapping) for Commerce pages (PDP, PLP) ensure that the server-side rendered response contains metadata that allows Google to differentiate pages.

Folder-mapped pages (without using [bulk metadata](https://www.aem.live/docs/bulk-metadata)) all look the same and Google might classify them as duplicate and not index them, despite different canonical URLs as Google considers canonical URLs a hint and not an absolute value.

### Multi-store setups

For multi-store setups or stores supporting multiple locales:

* Verify a [correct hreflang setup](https://www.aem.live/developer/sitemap#specifying-the-primary-language-manually). You can validate using the [hreflang Tags Testing Tool](https://technicalseo.com/tools/hreflang/)

* Verify that the store code is part of the URL for every Catalog Service or Live Search request. This is required because Google caches Catalog Service responses without considering headers. Not adding the store code to the URL might lead to Google indexing the wrong data.

  :::tip

>   You can ensure this by proxying the Catalog Service endpoint through Fastly. Alternatively, you can add a query parameter, but this method is incompatible with dropins.

Reviewer: Add link to the Fastly doc here?

  :::

### Redirects

Ensure proper redirects are set up using the redirects sheet or through your [CDN](/implementation/setup/content-delivery-network/).

> * Fallbacks for any potential URL changes (e.g. product URLs or missing .html suffix)

Reviewer: ```suggestion
* Fallbacks for any potential URL changes (for example, product URLs or missing .html suffixes)
```

| Document title            | `title`                                                                                                                                                                 |
| Meta tags                 | `description`, `keywords`, `og:type`, `og:title`, `og:description`, `og:url`, `og:image`, `og:image:secure_url`, `og:product:price:amount`, `og:product:price:currency` |
| Schema.org data (JSON-LD) | `WebSite`, `Product`, `AggregateRating`, `Rating`, `BreadcrumbList`                                                                                                     |

:::note
When importing pages, ensure that when a metadata block is present on a page, it does not contain any empty fields. This might lead to pages with missing titles, which Google penalizes.
:::

### Schema.org data (JSON-LD)

JavaScript Object Notation for Linked Data (JSON-LD) is a structured data format that helps search engines understand the content of your web pages more explicitly. It is typically added as a script tag in the `<head>` of your HTML document.

> Adobe Commerce storefronts on Edge Delivery Services should include schema.org annotations to expose product data to search engines. This data should be included on all Commerce pages, especially PDPs and PLPs.

Reviewer: ```suggestion
Adobe Commerce storefronts on Edge Delivery Services should include [schema.org](https://schema.org/) annotations to expose product data to search engines. This data should be included on all Commerce pages, especially PDPs and PLPs.
```

> You should compare the data available on the site before migrating to Edge Delivery Services to ensure that SKU's for all product variants are included. Use the [Rich Results Test](https://search.google.com/test/rich-results) to validate the schema.org annotations. The PDP dropin in the boilerplate contains an example for JSON-LD data.

Reviewer: ```suggestion
You should compare the data available on the site before migrating to Edge Delivery Services to ensure that SKUs for all product variants are included. Use the [Rich Results Test](https://search.google.com/test/rich-results) to validate the schema.org annotations. The PDP dropin in the boilerplate contains an example for JSON-LD data.
```

:::note
If you are using Google Shopping features (for example, products available on `shopping.google.com`), consider schema.org annotations with critical priority. If a single JSON-LD annotation on the page is invalid, Google considers all annotations invalid.
:::

### Generate metadata

> You can use the [PDP Metadata Generator](https://github.com/hlxsites/aem-boilerplate-commerce/tree/main/tools/pdp-metadata) tool to automate the process of generating all recommended metadata. The tool is designed to fetch product data from the Catalog Service, process it, and generate a metadata spreadsheet in `.xslx` format, which can be used for the [bulk metadata](https://www.aem.live/docs/bulk-metadata) feature in Edge Delivery Services.

Reviewer: ```suggestion
You can use the [PDP Metadata Generator](https://github.com/hlxsites/aem-boilerplate-commerce/tree/main/tools/pdp-metadata) tool to automate the process of generating all recommended metadata. The tool fetches product data from the Catalog Service, processes it, and generates a metadata spreadsheet in `.xslx` format. The spreadsheet can be used for the [bulk metadata](https://www.aem.live/docs/bulk-metadata) feature in Edge Delivery Services.
```

<Aside type="cool" title="Prerequisites">
* Node.js installed on your machine
* Access to the Catalog Service with the necessary API keys and configuration
</Aside>

To generate metadata using the PDP Metadata Generator tool:

<Steps>

> 1. Clone the boilerplate repository (if you haven't already).

Reviewer: This should be a prereq

1. Navigate to the `tools/pdp-metadata/` directory.
1. Install dependencies.
  
   ```bash
   npm install
   ```

   :::note
   Before running the tool, you must ensure that the `configFile` variable in the `pdp-metadata.js` file points to the correct configuration JSON file URL. This file contains the required parameters to access the Catalog Service API and should have been setup as part of your [project onboarding](/implementation/setup/commerce-configuration/).
   :::

> 1. Fetch product data, process it, and generate a file named `metadata.xlsx` in the project directory.

Reviewer: ```suggestion
1. Run the tool and generate a file named `metadata.xlsx` in the project directory.
```
